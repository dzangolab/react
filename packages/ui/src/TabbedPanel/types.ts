import type { ReactElement } from "react";

type TKeymap = {
  [key: string]: () => void;
};

type TOrientation = "horizontal" | "vertical";

type TPosition = "top" | "left" | "bottom" | "right";

type StorageType = "localStorage" | "sessionStorage";

type Tab = {
  label: string;
  children: React.ReactNode;
  key: string;
  icon?: string;
  closable?: boolean;
};

type Properties = {
  children?: ReactElement | ReactElement[];
  defaultActiveIndex?: number;
  id?: string;
  onTabChange?: (index: number) => void;
  position?: TPosition;
  persistState?: boolean;
  persistStateStorage?: StorageType;
  tabs?: Tab[];
};

export type { TKeymap, TOrientation, TPosition, Properties, StorageType, Tab };
