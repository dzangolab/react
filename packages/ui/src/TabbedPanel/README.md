# TabbedPanel component

## Installation

### npm

```
npm install @prefabs.tech/react-ui
```

### yarn

```
yarn add @prefabs.tech/react-ui
```

### pnpm

```
pnpm add @prefabs.tech/react-ui
```

## Usage

```
import React from "react";

import { SubPane, TabbedPanel } from "@prefabs.tech/react-ui";


const Page: React.FC<Properties> = ({ children }) => {
  return (
    <TabbedPanel>
      <SubPane
        icon="icon1.jpg"
        title="Topic 1">
        Pane 1 content
      </SubPane>
      <SubPane
        icon="icon2.jpg"
        title="Topic 2">
        Pane 2content
      </SubPane>
    </TabbedPanel>
  );
};

export default Page
```

## Props

### Accordion

| Prop     | Type                                   | Description                                           | Default | Required |
| -------- | -------------------------------------- | ----------------------------------------------------- | ------- | -------- |
| position | `top` \| `left` \| `bottom` \| `right` | The position of the tabs relative to the tabbed pane. | `top`   | No       |

### SubPane

| Prop  | Type     | Description                                     | Default | Required |
| ----- | -------- | ----------------------------------------------- | ------- | -------- |
| icon  | `string` | The icon to display in the subpane's title area | None    | No       |
| title | `string` | The text to display in the subpan'es title area | None    | Yes      |
