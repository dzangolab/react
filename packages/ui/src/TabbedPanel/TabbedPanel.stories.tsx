import "../../public/tab-theme.css";
import { Meta, StoryFn } from "@storybook/react";
import React from "react";
import { SubPane } from "../index";

import { TabbedPanel } from "./index";

type TabbedPanelType = typeof TabbedPanel;
type TabbedPanelProperties = Meta<TabbedPanelType>;

const TabbedPanelStory: TabbedPanelProperties = {
  title: "UI/TabbedPanel",
  component: TabbedPanel,
};

const Template: StoryFn<TabbedPanelType> = (arguments_) => (
  <TabbedPanel {...arguments_}>
    <SubPane icon="https://tinyurl.com/yw5ack2m" title="Topic 1">
      Pane 1 content
    </SubPane>
    <SubPane icon="https://tinyurl.com/5cjtm2y7" title="Topic 2">
      Pane 2content
    </SubPane>
  </TabbedPanel>
);

export const TopOrientation: TabbedPanelProperties = Template.bind({});
export const BottomOrientation: TabbedPanelProperties = Template.bind({});
export const LeftOrientation: TabbedPanelProperties = Template.bind({});
export const RightOrientation: TabbedPanelProperties = Template.bind({});

TopOrientation.args = {
  position: "top",
};

BottomOrientation.args = {
  position: "bottom",
};

LeftOrientation.args = {
  position: "left",
};

RightOrientation.args = {
  position: "right",
};

export default TabbedPanelStory;
