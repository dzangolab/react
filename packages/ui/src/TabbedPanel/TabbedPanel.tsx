import React, { useEffect, useMemo, useRef, useState } from "react";

import { getOrientation, onTabDown } from "./helper";

import type { Properties } from "./types";

const TabbedPanel: React.FC<Properties> = ({
  children,
  defaultActiveIndex = 0,
  position = "top",
  id = "",
  persistState = false,
}) => {
  const [active, setActive] = useState<number | null>(null);
  const tabReferences = useRef<(HTMLButtonElement | null)[]>([]);
  const childNodes = Array.isArray(children) ? children : [children];

  const activeTabStorageKey = useMemo(() => `tab-${id}-active-index`, [id]);

  useEffect(() => {
    if (persistState && id) {
      const activeTabIndex = localStorage.getItem(activeTabStorageKey);

      if (activeTabIndex !== null) {
        setActive(Number(activeTabIndex));
      } else {
        setActive(defaultActiveIndex);
      }
    } else {
      setActive(defaultActiveIndex);
    }
  }, [id, persistState, defaultActiveIndex]);

  useEffect(() => {
    if (persistState && id) {
      localStorage.setItem(activeTabStorageKey, String(active));
    }
  }, [active, id, persistState]);

  const handleFocus = (index: number) => {
    const tab = tabReferences.current[index];
    if (tab) {
      tab.focus();
    }
  };

  if (!children) {
    throw new Error("TabbedPanel needs at least one children");
  }

  if (active === null) {
    return null;
  }

  return (
    <div className={`tabbed-panel ${position}`}>
      <div role="tablist" aria-orientation={getOrientation(position)}>
        {childNodes.map((item, index) => {
          const isActive = active === index;
          const title = item.props.title;
          const icon = item.props.icon;
          const key = id ? `${id}-${index}` : index;

          return (
            <button
              onKeyDown={(event) => {
                onTabDown(
                  active,
                  event,
                  childNodes.length,
                  handleFocus,
                  getOrientation(position),
                );
              }}
              onFocus={() => setActive(index)}
              ref={(element) => (tabReferences.current[index] = element)}
              onClick={() => setActive(index)}
              key={key}
              role="tab"
              aria-label={title}
              aria-disabled={isActive}
              aria-selected={isActive}
              tabIndex={isActive ? 0 : -1}
              className={isActive ? "active" : ""}
            >
              {icon ? (
                <img src={icon} alt="title icon" aria-hidden="true" />
              ) : null}
              <span>{title}</span>
            </button>
          );
        })}
      </div>
      <div role="tabpanel">{childNodes[active]}</div>
    </div>
  );
};

export default TabbedPanel;
