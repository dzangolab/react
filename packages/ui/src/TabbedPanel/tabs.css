.tabbed-panel {
  display: flex;
  flex-direction: column;
  width: 100%;
}

.tabbed-panel.bottom {
  flex-direction: column-reverse;
}

.tabbed-panel > div[role="tablist"] {
  display: flex;
  flex-direction: column;
  list-style: none;
  height: max-content;
  padding: 0;
  margin: 0;
}

.tabbed-panel.bottom > div[role="tablist"],
.tabbed-panel.top > div[role="tablist"] {
  flex-direction: row;
  overflow-x: auto;
}

.tabbed-panel.left {
  flex-direction: row;
}

.tabbed-panel > div[role="tablist"] > button {
  width: max-content;
}

.tabbed-panel.left > div[role="tablist"] {
  align-items: flex-start;
}

.tabbed-panel.left > div[role="tablist"] > button.active {
  border-right: var(--tab-active-border);
}

.tabbed-panel.right {
  flex-direction: row-reverse;
}

.tabbed-panel.right > div[role="tablist"] > button.active {
  border-left: var(--tab-active-border);
}

.tabbed-panel.left > div[role="tablist"] > button,
.tabbed-panel.right > div[role="tablist"] > button {
  width: 100%;
}
