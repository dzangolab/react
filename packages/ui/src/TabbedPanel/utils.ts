import type { TKeymap, TOrientation, TPosition, StorageType } from "./types";
import type { KeyboardEvent } from "react";

const getOrientation = (position: TPosition) => {
  let orientation: TOrientation;

  switch (position) {
    case "top":
    case "bottom":
      orientation = "horizontal";
      break;

    case "left":
    case "right":
      orientation = "vertical";
      break;

    default:
      orientation = "horizontal";
      break;
  }

  return orientation;
};

const onTabDown = (
  active: number,
  event: KeyboardEvent<HTMLButtonElement>,
  count: number,
  handleFocus: (value: number) => void,
  orientation: TOrientation,
) => {
  const nextTab = () => handleFocus((active + 1) % count);
  const previousTab = () => handleFocus((active - 1 + count) % count);
  const firstTab = () => handleFocus(0);
  const lastTab = () => handleFocus(count - 1);
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  const doNothing = () => {};

  // https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/code
  const keyMap: TKeymap = {
    ArrowRight: orientation === "horizontal" ? nextTab : doNothing,
    ArrowDown: orientation === "vertical" ? nextTab : doNothing,
    ArrowLeft: orientation === "horizontal" ? previousTab : doNothing,
    ArrowUp: orientation === "vertical" ? previousTab : doNothing,
    Home: firstTab,
    End: lastTab,
  };

  const action = keyMap[event.key];
  if (action) {
    event.preventDefault();
    action();
  }
};

const getStorage = (storageType: StorageType): Storage => {
  return storageType === "sessionStorage" ? sessionStorage : localStorage;
};

const clearSavedTabState = (
  key: string,
  storageType: StorageType = "localStorage",
) => {
  const storage = getStorage(storageType);

  storage.removeItem(key);
};

export { getOrientation, onTabDown, clearSavedTabState, getStorage };
