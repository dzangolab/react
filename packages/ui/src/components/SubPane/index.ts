import SubPane from "./SubPane";

export { SubPane };
