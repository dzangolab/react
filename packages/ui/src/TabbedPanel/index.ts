import TabbedPanel from "./TabbedPanel";
import { clearSavedTabState } from "./utils";

export { TabbedPanel, clearSavedTabState };
