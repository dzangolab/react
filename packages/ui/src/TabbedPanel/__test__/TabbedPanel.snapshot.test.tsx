import { render } from "@testing-library/react";
import React from "react";
import { expect, test } from "vitest";

import { SubPane } from "../../components/SubPane";
import TabbedPanel from "../TabbedPanel";

test("Component matches snapshot", () => {
  const { container } = render(
    <TabbedPanel>
      <SubPane icon="icon1.jpg" title="Topic 1">
        Pane 1 content
      </SubPane>
      <SubPane icon="icon2.jpg" title="Topic 2">
        Pane 2 content
      </SubPane>
    </TabbedPanel>,
  );

  expect(container).toMatchSnapshot();
});
