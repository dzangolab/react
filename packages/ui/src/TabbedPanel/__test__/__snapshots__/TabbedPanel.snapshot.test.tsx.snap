// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Component matches snapshot 1`] = `
<div>
  <div
    class="tabbed-panel top"
  >
    <div
      aria-orientation="horizontal"
      role="tablist"
    >
      <button
        aria-disabled="true"
        aria-label="Topic 1"
        aria-selected="true"
        class="active"
        role="tab"
        tabindex="0"
      >
        <img
          alt="title icon"
          aria-hidden="true"
          src="icon1.jpg"
        />
        <span>
          Topic 1
        </span>
      </button>
      <button
        aria-disabled="false"
        aria-label="Topic 2"
        aria-selected="false"
        class=""
        role="tab"
        tabindex="-1"
      >
        <img
          alt="title icon"
          aria-hidden="true"
          src="icon2.jpg"
        />
        <span>
          Topic 2
        </span>
      </button>
    </div>
    <div
      role="tabpanel"
    >
      <div
        class="sub-pane"
      >
        Pane 1 content
      </div>
    </div>
  </div>
</div>
`;
